packages/controllers/src/subject-metadata/SubjectMetadataController.ts

```diff
@@ -0,0 +1,261 @@
+   */
+  clearState(): void {
+    this.update((_draftState) => {
+      return { ...defaultState };
+    });
+  }
+
+  /**
+   * Stores domain metadata for the given origin (subject). Deletes metadata for
+   * subjects without permissions in a FIFO manner once more than
+   * {@link SubjectMetadataController.subjectCacheLimit} distinct origins have
+   * been added since boot.
+   *
+   * In order to prevent a degraded user experience,
+   * metadata is never deleted for subjects with permissions, since metadata
+   * cannot yet be requested on demand.
+   *
+   * @param metadata - The subject metadata to store.
+   */
+  addSubjectMetadata(metadata: SubjectMetadataToAdd): void {
+    const { origin } = metadata;
+    const newMetadata: SubjectMetadata = {
+      ...metadata,
+      extensionId: metadata.extensionId || null,
+      host: metadata.host || null,
+      iconUrl: metadata.iconUrl || null,
+    };
+
+    let originToForget: string | null = null;
+    // We only delete the oldest encountered subject from the cache, again to
+    // ensure that the user's experience isn't degraded by missing icons etc.
+    if (this.subjectsEncounteredSinceStartup.size >= this.subjectCacheLimit) {
+      const cachedOrigin = this.subjectsEncounteredSinceStartup
+        .values()
+        .next().value;
+
+      this.subjectsEncounteredSinceStartup.delete(cachedOrigin);
+      if (!this.subjectHasPermissions(cachedOrigin)) {
+        originToForget = cachedOrigin;
+      }
+    }
+
+    this.subjectsEncounteredSinceStartup.add(origin);
+
+    if (
+      !newMetadata.extensionId &&
+      !newMetadata.host &&
+      /[a-z]+:\/\/\w.+/iu.test(origin)
+    ) {
+      try {
+        newMetadata.host = new URL(origin).host;
+      } catch (_) {
+        // Do nothing. Some "origins" are not valid URLs.
+      }
+    }
+
+    this.update((draftState) => {
+      // Typecast: ts(2589)
+      draftState.subjectMetadata[origin] = newMetadata as any;
+      if (typeof originToForget === 'string') {
+        delete draftState.subjectMetadata[originToForget];
+      }
+    });
+  }
+
+  /**
+   * Deletes all subjects without permissions from the controller's state.
+   */
+  trimMetadataState(): void {
+    this.update((draftState) => {
+      SubjectMetadataController.trimMetadataState(
+        // Typecast: ts(2589)
+        draftState as any,
+        this.subjectHasPermissions,
+      );
+    });
+  }
+
+  /**
+   * Deletes all subjects without permissions from the given state object. This
+   * method is static because we want to call it in the constructor, before the
+   * controller's state is initialized.
+   *
+   * @param draftState - The `immer` draft state.
+   * @param hasPermissions - A function that returns a boolean indicating
+   * whether a particular subject (identified by its origin) has any
+   * permissions.
+   */
+  private static trimMetadataState(
+    draftState: SubjectMetadataControllerState,
```

Comment:
The parameter is called `draftState` and it's documented as being an Immer draft, but the type contradicts this. It's typed as the state, not as a draft. It's also used this way in the constructor, where the state is passed in here directly (not as a draft).

Reply:
I think it would be simpler for this to return the trimmed state, rather than mutate the input. Then it doesn't matter whether what's passed in is a draft or not. We could then use `castImmutable` when passing in the draft, so that this doesn't need to accept a Draft type, letting us use this more easily in the constructor as well.

Reply:
Yeah, agreed. I still had to do an `any` cast due to the "infinite type" bug and therefore didn't use `castImmutable`, but I think this is better: ec6a216